### ISSUE TITLE
Windows: with sockconnect() Nvim does not fully exit

### ISSUE THREAD

**Issue by zdm** [neovim/neovim#14562]
Hi,
NVIM v0.5.0-dev+1325-gdab6b08a1
windows 10 x64

`nvim.exe` stays in memory after ui was closed, This happens only when `jsobstart()` was executed previously. I am starting long-running job (rpc service) then exit ui (`:qall!`). UI exits, job process is also killed, but `nvim.exe` left in processes list.
Why this is happens and how I can kill all related processes on exit?

**Comment by glacambre** [comment]
I believe I encountered this issue in Firenvim too.

**Comment by mjlbach** [comment]
I've only encountered this on Windows, does not seem to occur on mac/linux

**Comment by zdm** [comment]
I am not tested under linux.
Could somebody help to understand, how to close vim completely. 
Currently I have 100 processes in background after several hours of work.

**Comment by zdm** [comment]
Could somebody from the dev team help to solve this?
Neovim is not exited, it creates lot of background processes, which keeps files locked.

**Comment by glacambre** [comment]
@zdm I can't reproduce this issue but you can - so the logical thing to do here would be for you to investigate.

I think the first thing you will need to do is [build neovim with debug symbols](https://github.com/neovim/neovim/wiki/Building-Neovim#windows--msvc). Once that's done, you can try to reproduce the problem. When it occurs, attach to neovim with a debugger (I think [windbg](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/debugger-download-tools) can do that?) and look at the stack trace: this will tell us where neovim gets stuck and possibly why.

**Comment by zdm** [comment]
So, the problem is in the `sockconnect` call.

My workflow:
1. try to connect to the port 55555: `sockconnect("tcp", "127.0.0.1:55555", {"rpc": v:true})`;
2. if connection is refused, run service: `jobstart( "softvisio-cli.cmd rpc" )`;
3. connect again: `sockconnect("tcp", "127.0.0.1:55555", {"rpc": v:true})`;

If on stage 1 connection is refused - nvim does not exit normally when I close it.
This is only under windows, on linux nvim exits without problems.
I am using latest nightly build, currently `NVIM v0.5.0-dev+1449-gbdf3df402`.

**Comment by glacambre** [comment]
@zdm thank you, that's a useful piece of information. What does the call stack look like?

**Comment by zdm** [comment]
I didn't used debugger.
My code is called from nvim plugin. It has the function, that checks and creates the rpc channel.
Or what do you mean?

**Comment by glacambre** [comment]
What I mean is that you told us the steps that eventually cause neovim 
to get stuck, but we don't know where it gets stuck in the code. That's 
why we need to know what the call stack is like :).

**Comment by zdm** [comment]
I think failed `sockconnect` stays in event loop and prevents nvim to exit.

What do you mean under call stack? Sequence of the vimscript functions call? I didn't used debugger, so unable to provide calls stack on low level.

**Comment by zdm** [comment]
Can I help somehow to reproduce and fix this?
I can provide steps to reproduce using nodejs script as job.

**Comment by glacambre** [comment]
>Can I help somehow to reproduce and fix this?

In my opinion what would be most helpful would be to attach to the hung process with a debugger and to observe the call stack, to understand where Neovim is stuck :).

**Comment by zdm** [comment]
This is complex for me, I am not a C developer and never used such tools before.

**Comment by glacambre** [comment]
The wiki describes [how to build neovim](https://github.com/neovim/neovim/wiki/Building-Neovim#windows--msvc). Microsoft documentation details [how to download windbg](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/debugger-download-tools), [how to attach to a process](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/debugging-a-user-mode-process-using-windbg) and [how to view the call stack](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/calls-window). If you need help with a specific step feel free to ask questions, we'll be happy to provide as much help as we can :).

**Comment by zdm** [comment]
I am sorry, I am unable to use debugger, this is too complex for my small brains.

To reproduce it you can do following steps:

1. `:call sockconnect("tcp", "127.0.0.1:55552", {"rpc": v:true})`;
2. close neovim ui;
3. see, that ui closed but `nvim.exe` left in the processes list;

This is definitely the bug in the `sockconnect`.

**Comment by zdm** [comment]
I am surprised. Nobody cares at all. Could you at least put `bug` label to this issue?

**Comment by glacambre** [comment]
@zdm we need a call stack.

**Comment by zdm** [comment]
I understand, I am not able to provide it. I don't know, how this tools 
works, need to install and learn them. This is not my specialization. 
This is the same situation, if doctor instead of cure gives the 
instructions how to make operation by yourself.
I provided simple and clear instructions to reproduce this bug. Any 
neovim developer can get the required stack in a 5 minutes without any 
problems. Or it is important, that this stack must be provided by me 
exclusively? I didn't saw such requirements in the other neovim issues. 
;-) Could you please at least mark it as bug? Maybe somebody will have a 
time to fix it.

**Comment by jamessan** [comment]
> This is the same situation, if doctor instead of cure gives the instructions how to make operation by yourself.

I don't have a Windows system, so it would be difficult for me to get the required debug information.  Since you do, people are providing guidance on how you can provide that information, so we can have a better idea of how to fix the problem.  This is like a doctor asking you where it hurts.  It wouldn't be good if we operated on your leg when your arm was the problem.

> Or it is important, that this stack must be provided by me exclusively? I didn't saw such requirements in the other neovim issues.

No, it doesn't _need_ to be provided by you, but you're the one experiencing the issue.  If someone else is able to reproduce the issue, they could also provide the stack.  We don't really care _who_ provides the information, as long as someone does.

> Could you please at least mark it as bug?

@glacambre already added the label.  That's not going to change anything, though.  Labels don't make reproductions with relevant triage information appear.

**Comment by justinmk** [comment]
This is presumably a libuv ref hanging around, and loop_close waits forever.  https://github.com/neovim/neovim/blob/7e0c424aed118ab6d09f310b406ae877dbc1c141/src/nvim/event/loop.c#L133

**Comment by zdm** [comment]
Is it possible to fix this? I can help in testing.

**Comment by justinmk** [comment]
Busy with other things, happy to review a PR.

**Comment by erw7** [comment]
I have confirmed that the following steps have solved this problem with #15889.

```vim
:call sockconnect('tcp', '127.0.0.1:55552', {'rpc':v:true})
:call jobstart('nvim -u NORC', {'env':{'NVIM_LISTEN_ADDRESS':'127.0.0.1:55552'}})
:call sockconnect('tcp', '127.0.0.1:55552', {'rpc':v:true})
:q
tasklist | find "nvim"
```
